aesd-char-driver/main.c

```diff
@@ -0,0 +1,221 @@
+/**
+ * @file aesdchar.c
+ * @brief Functions and data related to the AESD char driver implementation
+ *
+ * Based on the implementation of the "scull" device driver, found in
+ * Linux Device Drivers example code.
+ *
+ * @author Dan Walkes
+ * @date 2019-10-22
+ * @copyright Copyright (c) 2019
+ *
+ * Few functions in this module are written by Vignesh Vadivel
+ *
+ */
+
+#include <linux/module.h>
+#include <linux/init.h>
+#include <linux/printk.h>
+#include <linux/types.h>
+#include <linux/cdev.h>
+#include <linux/slab.h>
+#include <linux/fs.h> // file_operations
+#include "aesdchar.h"
+int aesd_major = 0; // use dynamic major
+int aesd_minor = 0;
+
+MODULE_AUTHOR("Vignesh Vadivel");
+MODULE_LICENSE("Dual BSD/GPL");
+
+struct aesd_dev aesd_device;
+
+int aesd_open(struct inode *inode, struct file *filp){
+  PDEBUG("open");
+  filp->private_data = container_of(inode->i_cdev, struct aesd_dev, cdev);
+  return 0;
+}
+
+int aesd_release(struct inode *inode, struct file *filp){
+  PDEBUG("release");
+  return 0;
+}
+
+
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+                  loff_t *f_pos){
+  ssize_t bytesReadFromKernelSpace = 0;
+  ssize_t bytesUnread = 0;
+  struct aesd_buffer_entry *readDataNode; 
+  ssize_t bytesReadToUserSpace = 0;
+  int mutexLockReturn;
+
+  if (filp == NULL || buf == NULL || f_pos == NULL){
+    return -EFAULT;
+  }
+
+  PDEBUG("read %zu bytes with offset %lld", count, *f_pos);
+  
+  // Mutex Lock //
+  mutexLockReturn = mutex_lock_interruptible(&((struct aesd_dev *)filp->private_data)->lock);
```

Comment:
@Vignesh-Vadivel for readability you can define a local struct aesd_dev * and cast once at the start of the function